### ISSUE TITLE
[REVIEW]: Cylc: a Workflow Engine for Automating Cycling Systems

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#737]
**Submitting author:** @hjoliver (<a href="http://orcid.org/0000-0002-5715-5279">Hilary Oliver</a>)
**Repository:** <a href="https://github.com/cylc/cylc" target ="_blank">https://github.com/cylc/cylc</a>
**Version:** cylc-7.7.0
**Editor:** @arfon
**Reviewer:** @sarats, @Chilipp
**Archive:** <a href="https://doi.org/10.5281/zenodo.1251084" target="_blank">10.5281/zenodo.1251084</a>

## Status

[![status](http://joss.theoj.org/papers/ead20043fcabdb8fc28928d333429517/status.svg)](http://joss.theoj.org/papers/ead20043fcabdb8fc28928d333429517)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/ead20043fcabdb8fc28928d333429517"><img src="http://joss.theoj.org/papers/ead20043fcabdb8fc28928d333429517/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/ead20043fcabdb8fc28928d333429517/status.svg)](http://joss.theoj.org/papers/ead20043fcabdb8fc28928d333429517)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/cylc/cylc" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@sarats & @Chilipp, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.theoj.org/about#reviewer_guidelines. Any questions/concerns please let @arfon know.

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/cylc/cylc">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (cylc-7.7.0)?
- [x] **Authorship:** Has the submitting author (@hjoliver) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/cylc/cylc">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (cylc-7.6.1)?
- [x] **Authorship:** Has the submitting author (@hjoliver) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon. I'm here to help you with some common editorial tasks. @sarats, it looks like you're currently assigned as the reviewer for this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[--> Check article proof :page_facing_up: <--](https://github.com/openjournals/joss-papers/blob/joss.00737/joss.00737/10.21105.joss.00737.pdf)

**Comment by sarats** [comment]
@hjoliver Sure, let us know when it is ready for review.

**Comment by hjoliver** [comment]
@Chilipp @sarats - sorry for the delay, ready for review now as cylc-7.7.0.  I've updated the paper details on the cylc master branch; hopefully that flows through to the JOSS system here...

**Comment by Chilipp** [comment]
Hey @hjoliver thanks for your work on this! Unfortunately I am on vacation now for the next two weeks, but I am sure I can find the necessary time for the review next week on Tuesday or Wednesday.

**Comment by sarats** [comment]
@hjoliver Could you point me to the current version of the paper? 

Perhaps, you need to issue some commands to @whedon?
@whedon commands

**Comment by hjoliver** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List all of Whedon's capabilities
@whedon commands

# Assign a GitHub user as the sole reviewer of this submission
@whedon assign @username as reviewer

# Add a GitHub user to the reviewers of this submission
@whedon add @username as reviewer

# Remove a GitHub user from the reviewers of this submission
@whedon remove @username as reviewer

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

# Change editorial assignment
@whedon assign @username as editor

# Set the software archive DOI at the top of the issue e.g.
@whedon set 10.0000/zenodo.00000 as archive

# Open the review issue
@whedon start review

🚧 🚧 🚧 Experimental Whedon features 🚧 🚧 🚧

# Compile the paper
@whedon generate pdf

```

**Comment by hjoliver** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[--> Check article proof :page_facing_up: <--](https://github.com/openjournals/joss-papers/blob/joss.00737/joss.00737/10.21105.joss.00737.pdf)

**Comment by hjoliver** [comment]
@sarats - the PDF proof has updated.

**Comment by hjoliver** [comment]
@Chilipp - no worries, enjoy your vacation!

**Comment by Chilipp** [comment]
Hey @hjoliver ! Here are the first points concerning the *General checks*:

- The license file: 
   > **License:** Does the repository contain a plain-text LICENSE file with the contents of an OSI approved software license?

  You state in the [README](https://github.com/cylc/cylc/blob/master/README.md)
  > Cylc is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

  and 
  > You should have received a copy of the GNU General Public License along with cylc.

  but I could not find the LICENSE file in the repository.
- The version:
  > Version: Does the release version given match the GitHub release (cylc-7.6.1)?

  No, it's 7.7.0 now. So I am not sure, how to handle this (pinging @arfon)

**Comment by Chilipp** [comment]
Hey @hjoliver , I tried your [Workflow example](https://cylc.github.io/cylc/documentation.html#a-cycling-workflow-example) and it works well (including installation) until I try to run the job. Then it stays in the *submitted* state for the first job.

![image](https://user-images.githubusercontent.com/9960249/40990821-45adf0a4-68f2-11e8-9c3b-8a3e60bb86a0.png)

I have to say, I tested it on my laptop (OSX) without a scheduler. But from what I got in the documentation, that means that I use the *background* scheduler by default which just invokes a bash background job.
So what am I doing wrong here?

**Comment by hjoliver** [comment]
@Chilipp - yes, cylc can manage background jobs, even on remote hosts.  I'll respond regarding OSX at https://github.com/cylc/cylc/issues/2689

**Comment by Chilipp** [comment]
Dear all,

I am done now with my review so far, went through the user guide and thanks a lot for this nice and well-developed open source package. I find it well documented and helpful and therefore, if the points already mentioned previously (version https://github.com/openjournals/joss-reviews/issues/737#issuecomment-394667300 and contribution guide https://github.com/openjournals/joss-reviews/issues/737#issuecomment-395481912) are fixed and commented by @arfon, I recommend to accept the paper.

Personally, however, there are two points I would like to mention:

1. I am not happy about the reliance on Python2 and GTK, but as you already state in the paper, this is already on the roadmap and this is anyway no reason for JOSS to reject the paper.
2. Concerning the tests on [Travis CI](https://travis-ci.org/cylc/cylc) I wonder, how you are tracking how well your software is covered by the tests? The core of it is a python module, so it should be possible to some extent at least using [coverage](https://coverage.readthedocs.io/en/coverage-4.5.1/) or some equivalent tool.
   It would be very helpful (for reviewing and for you as a developer) to have a coverage report. But of course, due to the complexity and scale of the software, I understand that this is not completely straight-forward and anyway, the questions above just ask
   > Are there automated tests or manual steps described so that the function of the software can be verified?

   So this is definitely fulfilled and, additionally, it seems to me that your test suite covers all the central parts of the software, tutorials, etc.. Therefore I recommend to ignore this shortcoming.

**Comment by hjoliver** [comment]
@sarats - thank you. In reponse to your points:
 * LICENSE vs COPYING - let's see the response from @arfon 
 * ditto - need advice from @arfon on how to correct the version in the submission
 * lack of CONTRIBUTING.md is a problem - we have planned to correct this since some months ago, but still waiting on legal advice on that from my home institution (I will chase them up...)
 * we have Tracis CI hooked up to the main cylc repository on GitHub - every new commit triggers an increasingly massive battery of tests (currently takes almost 50 minutes to run, typically). Similar Codacy for static source code analysis. Code coverage testing is yet to be investigated and tested, but it is on my radar.

**Comment by arfon** [comment]
> - LICENSE info is in COPYING file as per GPL v3 whereas the Journal requires a LICENSE file? @arfon Can you clarify?

`COPYING` is acceptable, especially as this is recommended by the GPL.

> - Version needs to be fixed in submission as pointed by @Chilipp.

I can update this. What should the new version be?

**Comment by hjoliver** [comment]
Thanks @arfon 

New version (as reviewed) is 7.7.0

**Comment by sarats** [comment]
> * lack of CONTRIBUTING.md is a problem - we have planned to correct this since some months ago, but still waiting on legal advice on that from my home institution (I will chase them up...)

@hjoliver I understand how complex formulating this policy can be. I'm involved a U.S. Dept. of Energy funded multi-lab project where we encourage people willing to contribute to submit bug reports or fixes for simple scenarios.  https://github.com/E3SM-Project/E3SM/blob/master/CONTRIBUTING.md
However we recommend a formal collaboration plan for more extensive changes. 

Food for thought: You can perhaps have something simple for the purpose of this review before finalizing your policy for extensive IP transfer etc.? 
Anyway, this is the only blocker. 

Otherwise this submission is good to go. Thanks for all the hard work.

**Comment by Chilipp** [comment]
> Otherwise this submission is good to go. Thanks for all the hard work.

The same for me. Thanks!

**Comment by hjoliver** [comment]
Thanks very much @sarats and @Chilipp (and @arfon).  I'll discuss with the team in the next few days what we can do about a preliminary contributing.md ...

**Comment by hjoliver** [comment]
(I've been given the legal go-ahead for a contributor agreement, will try to add something up this week).

**Comment by arfon** [comment]
How are you getting along here @hjoliver?

**Comment by hjoliver** [comment]
I've just posted Pull Request to Cylc, with a new contributor license agreement. We should be good to go when that's reviewed and merged to master ... I'll report back to confirm shortly.

**Comment by hjoliver** [comment]
@arfon - new `CONTRIBUTING.md` CLA just merged to Cylc master: https://github.com/cylc/cylc/blob/master/CONTRIBUTING.md

**Comment by sarats** [comment]
Looks good to go.

**Comment by Chilipp** [comment]
Hey @hjoliver  and @arfon! Sorry for the delay, it looks good to me, too 👍

**Comment by arfon** [comment]
@hjoliver - At this point could you make an archive of the reviewed software in Zenodo/figshare/other service and update this thread with the DOI of the archive? I can then move forward with accepting the submission.

**Comment by hjoliver** [comment]
@arfon - the Zenodo DOI for cylc-7.7.0 (as reviewed) is: [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.1251084.svg)](https://doi.org/10.5281/zenodo.1251084)

**Comment by arfon** [comment]
@whedon set 10.5281/zenodo.1251084 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.1251084" target="_blank">10.5281/zenodo.1251084</a> is the archive.

**Comment by arfon** [comment]
@sarats, @Chilipp - many thanks for your reviews here ✨ 

@hjoliver - your paper is now accepted into JOSS and your DOI is https://doi.org/10.21105/joss.00737 :zap: :rocket: :boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](http://joss.theoj.org/papers/10.21105/joss.00737/status.svg)](https://doi.org/10.21105/joss.00737)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.00737">
  <img src="http://joss.theoj.org/papers/10.21105/joss.00737/status.svg" alt="DOI badge" >
</a>
```
This is how it will look in your documentation:

[![DOI](http://joss.theoj.org/papers/10.21105/joss.00737/status.svg)](https://doi.org/10.21105/joss.00737)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: http://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://www.flipcause.com/secure/cause_pdetails/Mjk3Nzk=

**Comment by hjoliver** [comment]
@arfon @sarats @Chilipp - that's great, thanks very much :grin:
